kore/app/rpc/Main.hs

```diff
@@ -2,8 +2,172 @@
+        <*> parseKoreSolverOptions
+        <*> parseKoreLogOptions Main.exeName startTime
+        <*> parseBugReportOption
+        <*> option
+            auto
+            ( metavar "SERVER_PORT"
+                <> long "server-port"
+                <> help "Port for the RPC server to bind to"
+            )
+  where
+    parseMainModuleName =
+        GlobalMain.parseModuleName
+            "MODULE"
+            "module"
+            "The name of the main module in the Kore definition."
+
+-- | modifiers for the Command line parser description
+parserInfoModifiers :: InfoMod options
+parserInfoModifiers =
+    fullDesc
+        <> header "kore-rpc - a JSON RPC server for symbolically executing Kore definitions"
+
+exeName :: GlobalMain.ExeName
+exeName = GlobalMain.ExeName "kore-rpc"
+
+-- | Environment variable name for extra arguments
+envName :: String
+envName = "KORE_RPC_OPTS"
 
 main :: IO ()
-main = runServer 31337
+main = do
+    startTime <- getTime Monotonic
+    options <-
+        GlobalMain.mainGlobal
+            Main.exeName
+            (Just envName)
+            (parseKoreRpcServerOptions startTime)
+            parserInfoModifiers
+    for_ (GlobalMain.localOptions options) mainWithOptions
+
+mainWithOptions :: GlobalMain.LocalOptions KoreRpcServerOptions -> IO ()
+mainWithOptions localOptions@GlobalMain.LocalOptions{execOptions = KoreRpcServerOptions{koreSolverOptions, koreLogOptions, bugReportOption}} = do
+    ensureSmtPreludeExists koreSolverOptions
+    exitWith
+        =<< withBugReport
+            Main.exeName
+            bugReportOption
+            ( \tmpDir ->
+                koreRpcServerRun localOptions
+                    & handle handleSomeException
+                    & runKoreLog
+                        tmpDir
+                        koreLogOptions
+            )
+
+koreRpcServerRun ::
+    GlobalMain.LocalOptions KoreRpcServerOptions ->
+    GlobalMain.Main ExitCode
+koreRpcServerRun GlobalMain.LocalOptions{execOptions, simplifierx} = do
+    GlobalMain.SerializedDefinition{serializedModule, lemmas} <-
+        GlobalMain.deserializeDefinition
+            simplifierx
+            koreSolverOptions
+            definitionFileName
+            mainModuleName
+    let SerializedModule{metadataTools} = serializedModule
+    let smtConfig =
+            SMT.defaultConfig
+                { SMT.timeOut = timeOut
+                , SMT.rLimit = rLimit
+                , SMT.resetInterval = resetInterval
+                , SMT.prelude = prelude
+                }
+
+    -- launch the SMT solver, initialize it and then pass the SolverSetup object to the rpc server
+    GlobalMain.clockSomethingIO "Executing" $
+        bracket
+            (SMT.newSolver smtConfig)
+            SMT.stopSolver
+            $ \mvar -> do
+                let solverSetup = SMT.SolverSetup{userInit = SMT.runWithSolver $ declareSMTLemmas metadataTools lemmas, refSolverHandle = mvar, config = smtConfig}
```

Comment:
Could this line be split up somehow? e.g. above the call to `clockSomethingIO` add a new binding for `setupWithMVar :: MVar ? -> SolverSetup`?